main.py

```diff
@@ -2303,40 +2309,69 @@ def on_modified_async(self):
+                if self._visible:
                     self.view.hide_popup()
 
     def handle_response(self, response, point):
         if response is not None:
-            config = config_for_scope(self.view)
-            signatures = response.get("signatures")
-            activeSignature = response.get("activeSignature")
-            debug("got signatures, active is", len(signatures), activeSignature)
-            if len(signatures) > 0 and config:
-                signature = signatures[activeSignature]
-                debug("active signature", signature)
-                formatted = []
-                formatted.append(
-                    "```{}\n{}\n```".format(config.languageId, signature.get('label')))
-                params = signature.get('parameters')
-                if params:
-                    for parameter in params:
-                        paramDocs = parameter.get('documentation')
-                        if paramDocs:
-                            formatted.append("**{}**\n".format(parameter.get('label')))
-                            formatted.append("* *{}*\n".format(paramDocs))
-
-                formatted.append(signature.get('documentation'))
-
-                mdpopups.show_popup(
-                    self.view,
-                    preserve_whitespace("\n".join(formatted)),
-                    css=".mdpopups .lsp_signature { margin: 4px; } .mdpopups p { margin: 0.1rem; }",
-                    md=True,
-                    flags=sublime.HIDE_ON_MOUSE_MOVE_AWAY,
-                    location=point,
-                    wrapper_class="lsp_signature",
-                    max_width=800)
+            self._config = config_for_scope(self.view)
+            self.signatures = response.get("signatures")
+            self.active_signature = response.get("activeSignature", -1)
+            debug("got signatures, active is", len(self.signatures), self.active_signature)
+            if len(self.signatures) > 0 and self._config:
+                mdpopups.show_popup(self.view,
+                                    self._build_popup_content(),
+                                    css=self.__class__.css,
+                                    md=True,
+                                    flags=sublime.HIDE_ON_MOUSE_MOVE_AWAY,
+                                    location=point,
+                                    wrapper_class=self.__class__.wrapper_class,
+                                    max_width=800,
+                                    on_hide=self._on_hide)
+                self._visible = True
+
+    def on_query_context(self, key, _, operand, __):
+        if key != "lsp.signature_help":
+            return False  # Let someone else handle this keybinding.
+        elif not self._visible:
+            return False  # Let someone else handle this keybinding.
+        elif self.active_signature not in range(0, len(self.signatures)):
+            return False  # Let someone else handle this keybinding.
```

Comment:
What happens if we just check for `-1` here?

Reply:
When there are zero signatures, you might get an `activeSignature` of `0` from a server, and a `signatures == []`. In that case, `signatures[active_signature]` will raise an out of bounds exceptions. I like this approach of checking the range because it accounts for both invalid numbers for the active signature as well as for an empty signature list.

Reply:
We should guard against this case in `handle_response` and at least log an error.
See https://github.com/Microsoft/vscode-languageserver-node/issues/191 for a discussion with one of the protocol maintainers about defaulting to 'null' instead of '0'

Reply:
I see. I should fix that in clangd then 💃. For the time being, I think checking if the active signature is not out-of-bounds is a good safe-guard for faulty servers.